enhancements/windows-containers/health-management.md

```diff
@@ -0,0 +1,258 @@
+start the Kubelet Windows Service.
+
+In order to accomplish the goals listed in this enhancement, I am proposing that WMCB be converted into a daemon,
+Windows Machine Config Daemon (WMCD). WMCD will have the following responsibilities:
+* WMCD maintains current responsibilities of bootstrapping kubelet through the `bootstrap` command.
+* Using the configuration provided by a ConfigMap created by WMCO, WMCD maintains the state of Windows Services on the VM.
+* WMCD cleans up all changes it has made to a VM when ran with the `deconfigure` command.
+
+The Windows Node creation workflow for the end user should not be changed.
+
+### Services ConfigMap
+
+WMCO will create and maintain an immutable ConfigMap with the naming scheme:
+`services-<MajorVersion>-<MinorVersion>-<PatchVersion>-<Commit>`. For example, if WMCO was built from commit a7b5
+of WMCO version 5.0.0, the created ConfigMap would be named `config-5-0-0-a7b5`.
+
+The purpose of this ConfigMap is to provide WMCD with the specifications required to configure a Node. The ConfigMap
+will be watched via the existing ConfigMap controller. If it is created with incorrect values, WMCO will delete it.
+If it is deleted, WMCO will re-create it. If it is not present when WMCO is kicking off the configuration of a VM,
+it will be created.
+
+The services ConfigMap contains two keys: `services`, and `files`. The values for both keys are JSON objects.
+The services key contains all data required to configure the required Windows services on any VM that is to be
+added to the cluster as a Node. The proposed schema is as follows:
+```json
+[
+  {
+    "name": "name of the Windows service",
+    "command": "the command that will be ran",
+    "nodeVariablesToSubstitute": {
+      "name": "string within command field that will be substituted",
+      "jsonPathNodeObject": "value that will be substituted, taken from the provided jsonPath of the Node object"
+    },
+    "powershellVariablesToSubstitute": {
+      "name": "string within command field that will be substituted",
+      "path": "value that will be substituted, taken from the result of the powershell script at this path"
+    }
+  }
+]
+```
+
+The files key contains the path and checksum of files copied to the VM by WMCO. This will be used to validate that
+essential files have not been tampered with.
+```json
+[
+  {
+    "path": "filepath",
+    "checksum": "checksum of file"
+  }
+]
+```
+
+### Node configuration: WMCO responsibilities
+
+1) For a given Machine or BYOH VM, WMCO copies the payload and WMCD binary to the VM.
+2) WMCO runs WMCD with the bootstrap command, this results in the node object being created.
+3) WMCO annotates the Node object with the `desiredVersion` annotation.
+4) The value of this is set to the current WMCO version.
+5) WMCO starts WMCD as a service.
+6) Reconciliation can successfully return at this point.
```

Comment:
Are you sure we want to do this? How do we authoritatively inform the user that an instance has not been configured?

Reply:
>At this point WMCO has done everything required of it so it seems right for it to finish reconciliation.
It can ssh into the instance, correctly copy files, has created the Node, and successfully started the daemon.
Re-queuing should not cause a change in state.

How does it create the Node? Doesn't that happen as a result of the kubelet running?

>This would impact our ability to scale, with the trade off that the error would be logged to WMCO's logs. At the moment I'm not convinced that that's worth it.

We shouldn't worry about scaling in this enhancement as improving scaling is not one of the end goals.

My concern is that if we are doing a fire and forget approach, how do we guarantee that only one node is going to be upgraded at one time like we have today?

Reply:
I don't follow how you are going to be upgrading one node at a time with this fire and forget approach.

Reply:
Added more detail in the upgrage section
```
 3) Update the desired version annotation on the Node, to the current WMCO version. This should be done one Node as a
-   time, in order to ensure that workloads can continue to run with minimal interruption.
+   time, in order to ensure that workloads can continue to run with minimal interruption. WMCO is okay to change a
+   Node's desired version annotation if: there are zero Nodes with a mismatched desired and current version; there
+   are zero NotReady Nodes which have a current version matching WMCO's version.
```